hummingbot/connector/budget_checker.py

```diff
@@ -0,0 +1,290 @@
+        Yet others can charge the fees for both buy and sell orders in the quote asset.
+        """
+        if order_candidate.order_side == TradeType.BUY:
+            trading_pair = order_candidate.trading_pair
+            price = order_candidate.price
+            amount = order_candidate.amount
+            base, quote = split_hb_trading_pair(trading_pair)
+            fee = self._exchange.get_fee(
+                base, quote, order_candidate.order_type, order_candidate.order_side, amount, price
+            )
+            fee_adjustment = fee.fee_amount_in_quote(trading_pair, price, amount)
+        else:
+            fee_adjustment = Decimal("0")
+        return fee_adjustment
+
+    def _lock_available_collateral(self, order_candidate: OrderCandidate):
+        self._locked_collateral[order_candidate.collateral_token] += order_candidate.collateral_amount
+
+    def _adjust_amounts(
+        self, order_candidate: OrderCandidate, available_collateral: Decimal
+    ) -> OrderCandidate:
+        if order_candidate.order_side == TradeType.BUY:
+            trading_pair = order_candidate.trading_pair
+            base, quote = split_hb_trading_pair(trading_pair)
+            fee = self._exchange.get_fee(
+                base,
+                quote,
+                order_candidate.order_type,
+                order_candidate.order_side,
+                order_candidate.amount,
+                order_candidate.price,
+            )
+            adjusted_amount = fee.order_amount_from_quote_with_fee(
+                trading_pair, order_candidate.price, available_collateral
+            )
+            adjusted_amount = self._exchange.quantize_order_amount(trading_pair, adjusted_amount)
+        else:
+            adjusted_amount = available_collateral / order_candidate.price
+
+        adjusted_candidate = copy(order_candidate)
+        adjusted_candidate.amount = adjusted_amount
+        adjusted_candidate.collateral_amount = available_collateral
+
+        return adjusted_candidate
+
+
+class PerpetualBudgetChecker(BudgetChecker):
+    def __init__(self, exchange: exchange_base.ExchangeBase):
+        """
+        In the case of derived instruments, the collateral can be any token.
+        To get this information, this class uses the `get_buy_collateral_token`
+        and `get_sell_collateral_token` methods provided by the `PerpetualTrading` interface.
+        """
+        super().__init__(exchange)
+        self._validate_perpetual_connector()
+
+    def populate_collateral_fields(self, order_candidate: OrderCandidate) -> OrderCandidate:
+        trading_pair = order_candidate.trading_pair
+
+        if order_candidate.order_side == TradeType.BUY:
+            collateral_token = await self._exchange.get_buy_collateral_token(trading_pair)
+        else:
+            collateral_token = await self._exchange.get_sell_collateral_token(trading_pair)
+
+        order_size = order_candidate.amount * order_candidate.price
```

Comment:
The size to consider for the collateral will depend on the order type and the token the collateral should be provided in. It will not always be the same calculation.